### ISSUE TITLE
Jodd versioning

### ISSUE THREAD

**Issue by neroux** [oblac/jodd#388]
Is there currently a particular system behind Jodd's versioning?

Judging from the three-level versions one could assume it follow [semantic versioning](http://semver.org/), but that doesnt seem to be the case.

This issue is not to suggest semantic versioning (it has become quite popular in recent years, but software did great before it too) but I'd just like to start a discussion whether Jodd should have a versioning concept (assuming there isnt one yet). If so, I guess the next step would be to decide on one (-> Jodd XP? ;-) )

Semver might be certainly an option, but considering that Jodd's codebase is quite dynamic (to avoid saying "agile" :) ) its major version would get bumped quite often and we'd be now probably at Jodd 23.x

**Comment by pmouawad** [comment]
Hello,
If I can add my 2 cents to this.
As a user of JODD (log, lagarto and core) within JMeter , I am very happy but really frustrated of lack of visibility on release contents unless I missed something.
- http://jodd.org/release.html does not show anything unless I missed something.
- There is no milestone in Github, except 3.8.5 ? why 3.8.5 and not 3.8.2 or 3.9 ?

This is a bit annoying as you don't know anything about backward compatibility and why you should upgrade .
It should be your marketing page to show all the great new features no ?
Am I missing something ?
Thanks

**Comment by neroux** [comment]
Agreed, the changelog stopped with 3.6.7 and it being missing is an issue. Though a different one than versioning :)

**Comment by igorspasic** [comment]
@neroux @pmouawad you are both very right - and I apology for this mess :( To be honest, last few releases were made faster then usual, so to push the changes for different users who needed them and lack of time was the only reason why we haven't release it properly. So far the versioning was done without any strong rule; it was loose semver I would say.

I am perfectly fine to make the next release 3.9 or 4.0 and start doing it properly :)))

The only issue I see is the modules. So far I keep version unique across all the libraries. This _really_ make things easier for dependency management. I guess that this situation would break semver contract:

+ module A is not changed, version 1.1
+ module B significantly changed, requiring version 2.0

So what to do in this case?  Should we:

+ update module A to 2.0 because of B, although it does not depend on it; and there are no changes
+ update just modules B version to 2.0

I would go with the first as managing versions for 15 modules may be... well, though :)

**Comment by neroux** [comment]
I would tend to the second option. Versions could become meaningless if there is a mass bump just because one module had major rework. What would be your concern in regard to managing the version numbers? It would require individual version upgrades if they are justified for a module, but thats the point of it, isnt it?

**Comment by pmouawad** [comment]
Hi,
How do other similar proceed?

As a user, the first option is easier for me.
But I agree with @neroux opinion.

Regards

**Comment by pmouawad** [comment]
But second option means clear release notes and how will your proceed for issues on github ? one milestone id per lib?

**Comment by igorspasic** [comment]
I am ok with the steady major version increments; Jodd was always like that.

However, to be completely honest: I am bit afraid to go with separate versions at the moment...  I am aware this is the proper way to do it, but this might give more work on the infrastructure that I simply might not be able to support. In fact, if we go this way, the proper way would be to split Jodd repo into multiple repos, and work on each completely individually - right?

So how about this: to put this to backlog for version 3.x. I would really like to upgrade documentation and to work more on examples instead, many things there are pending. Then for version 4 (that can be later this year) we can really split Jodd into multiple, separate repos, and progress with each one _individually_.

Sounds good?

**Comment by neroux** [comment]
I mentioned the major version because I has stayed on 3.x now for the past eight years (pre-3 is not even mentioned) and if semantic versioning was to be properly implemented we'd jumped up a couple of versions only in the past few weeks (the refactoring of the email classes for example) and Jodd could quickly reach. That would be simply my concern about Semver.

I quickly checked and couldnt find specific version information for the individual modules.

**Comment by igorspasic** [comment]
That us exactly my concern for semver on global level. Thats why I guess we should split into more repos.

**Comment by neroux** [comment]
So you suggest to version only the individual module in that scenario, but not the full package (like now, _Jodd 3.x_) anymore, right?

That would be a solution, it would probably not even require separate repositories.

**Comment by igorspasic** [comment]
yes @neroux

+ If we add proper semver, then versions of the modules would be separate. True, this can be done without repository split.
+ The 'global' version of Jodd as it is now is gone and replaced with e.g. `1702` (based on date) Note that we still need to build a bundle version, as ppl are still using it :)

**Comment by neroux** [comment]
Sounds like a fair approach :+1:

**Comment by moh-sushi** [comment]
my 2 cents:
I see jodd as "one library" that is all modules should have the same version.


case "modules have separate versions":
Does every module have its own release procedure?
After release of module x , does a new global jodd version will be released, too?
If not, i have to exclude module x in pom.xml and add explicit dependency for module x? correct?

**Comment by igorspasic** [comment]
@moh-sushi yes, that is always been my concern, so to answer the questions:

+ every module would have its own release procedure and versioning.
+ After single release of module X, Jodd global version would not change, until certain amount of changes get collected - the same it is now.
+ There is always going to be a latest bundle pom, with versioning going up crazy (probably using date stamp).

I guess at the end you will have the same behavior as before, except you can track modules individually if you wish.

In other words, you would have two versioning:

+ each modules has its own versioning
+ bundle == Jodd version == date stamp

Something like that... but again, have to think it through one more time :)

**Comment by igorspasic** [comment]
btw, hope you like the [changelog](https://github.com/oblac/jodd/blob/master/CHANGELOG.md)

**Comment by pmouawad** [comment]
I love it :-) Thanks !

**Comment by pmouawad** [comment]
Hi @igr ,
I wonder if you know about this :
http://www.yegor256.com/2014/08/19/how-to-release-to-maven-central.html

I once read you were annoyed by release process and artifacts publication.

Once implemented on your project, release will be a matter of one chat with a bot :
"@rultor release, tag is `X.X.X`" :-)

**Comment by neroux** [comment]
@pmouawad How would that fix the versioning of the different modules and Jodd as a whole?

**Comment by pmouawad** [comment]
It would not, it was just a note related to release process. Maybe not the correct issue for that comment.

**Comment by igr** [comment]
Hello, everyone!

I've decided to postpone this decision. Simply put, don't have enough capacity to proceed with this now.

Please note that I'm not saying that you are not right - you are, and I also don't like bumping version on the code that is not changed - but at the moment, it is incredibly easy to maintain just one number.

The question about versioning goes bit further. As you know, Jodd contains a bunch of frameworks and when talking about versioning I can't stop wondering if it makes sense to have it all. Maybe it's time just to split into 3-4 libraries. Maybe not. Anyway, that is a different story.

I know this doesn't sound too cool, but just to save my sanity, I will keep versioning as it was before. Meanwhile, I will put the effort in properly managing the project on GitHub, writing release notes and to upgrade the documentation (much needed). This is a _temporal_ decision and I hope things will change in the future and we will find a way to organize things better.

I hope you understand.
